src/placeos-rest-api/controllers/edges.cr

```diff
@@ -0,0 +1,233 @@
+require "hound-dog"
+require "placeos-core/client"
+require "promise"
+
+require "./application"
+require "./systems"
+
+module PlaceOS::Api
+  class Edges < Application
+    base "/api/engine/v2/edges/"
+
+    before_action :check_admin, except: [:index, :show, :edge]
+    before_action :check_support, only: [:index, :show]
+
+    before_action :current_edge, only: [:destroy, :drivers, :show, :update, :update_alt]
+
+    skip_action :authorize!, only: [:edge]
+    skip_action :set_user_id, only: [:edge]
+    skip_action :check_oauth_scope, only: [:edge]
+
+    getter current_edge : Model::Edge { find_edge }
+
+    class_getter connection_manager : ConnectionManager { ConnectionManager.new(core_discovery) }
+
+    # Validate the present of the id and check the secret before routing to core
+    ws("/control", :edge) do |socket|
+      token = params["token"]?
+
+      render status: :unprocessable_entity, json: {error: "missing 'token' param"} if token.nil? || token.presence.nil?
+
+      edge_id = Model::Edge.validate_token(token)
+      head status: :unauthorized if edge_id.nil?
+
+      Log.info { {edge_id: edge_id, message: "new edge connection"} }
+
+      Edges.connection_manager.add_edge(edge_id, socket)
+    end
+
+    get("/:id/token", :token) do
+      head :unauthorized unless is_admin?
+      render json: {token: current_edge.token(current_user)}
+    end
+
+    def index
+      elastic = Model::Edge.elastic
+      query = elastic.query(params)
+      query.sort(NAME_SORT_ASC)
+      render json: paginate_results(elastic, query)
+    end
+
+    def show
+      render json: current_edge
+    end
+
+    def update
+      edge = current_edge
+      edge.assign_attributes_from_json(request.body.as(IO))
```

Comment:
400 if `request.body.nil?` then pass the non-nilable type rather than casting.

Reply:
via an intermediate assignment, yep